File: winui/CellposeGUI/Services/ImageLoaderService.cs
using CellposeGUI.Models;

namespace CellposeGUI.Services;

public sealed class ImageLoaderService
{
    public ImageData Load(string path)
    {
        var extension = Path.GetExtension(path).ToLowerInvariant();
        return extension switch
        {
            ".tif" or ".tiff" => LoadTiff(path),
            ".png" or ".jpg" or ".jpeg" or ".gif" or ".bmp" => LoadRaster(path),
            _ => throw new SidecarException($"Unsupported image format: {extension}"),
        };
    }

    private static ImageData LoadRaster(string path)
    {
        using var image = SixLabors.ImageSharp.Image.Load<SixLabors.ImageSharp.PixelFormats.Rgb24>(path);
        var width = image.Width;
        var height = image.Height;
        var pixels = new byte[width * height * 3];
        image.CopyPixelDataTo(pixels);
        return new ImageData
        {
            Width = width,
            Height = height,
            Channels = 3,
            Pixels = pixels,
        };
    }

    private static ImageData LoadTiff(string path)
    {
        using var tiff = BitMiracle.LibTiff.Classic.Tiff.Open(path, "r");
        if (tiff == null)
            throw new SidecarException($"Failed to open TIFF: {path}");

        var width = tiff.GetField(BitMiracle.LibTiff.Classic.TiffTag.IMAGEWIDTH)[0].ToInt();
        var height = tiff.GetField(BitMiracle.LibTiff.Classic.TiffTag.IMAGELENGTH)[0].ToInt();
        var samplesPerPixel = tiff.GetField(BitMiracle.LibTiff.Classic.TiffTag.SAMPLESPERPIXEL)?[0].ToInt() ?? 1;
        var channels = Math.Clamp(samplesPerPixel, 1, 3);
        var raster = new byte[width * height * channels];
        var scanline = new byte[tiff.ScanlineSize()];

        for (var y = 0; y < height; y++)
        {
            if (!tiff.ReadScanline(scanline, y))
                throw new SidecarException($"Failed to read TIFF scanline {y}");

            if (channels == 1)
            {
                Buffer.BlockCopy(scanline, 0, raster, y * width, width);
                continue;
            }

            for (var x = 0; x < width; x++)
            {
                var dst = (y * width + x) * channels;
                for (var c = 0; c < channels; c++)
                    raster[dst + c] = scanline[x * samplesPerPixel + c];
            }
        }

        return new ImageData
        {
            Width = width,
            Height = height,
            Channels = channels,
            Pixels = raster,
        };
    }
}

File: swiftui/CellposeGUI/Services/ImageLoaderService.swift
import CoreGraphics
import CellposeGUICore
import Foundation
import ImageIO

final class ImageLoaderService {
    func load(path: String) throws -> ImageData {
        let ext = URL(fileURLWithPath: path).pathExtension.lowercased()
        switch ext {
        case "tif", "tiff":
            return try loadTiff(path: path)
        case "png", "jpg", "jpeg", "gif", "bmp":
            return try loadRaster(path: path)
        default:
            throw SidecarError.serverError("Unsupported image format: .\(ext)")
        }
    }

    private func loadRaster(path: String) throws -> ImageData {
        let url = URL(fileURLWithPath: path)
        guard let source = CGImageSourceCreateWithURL(url as CFURL, nil),
              let cgImage = CGImageSourceCreateImageAtIndex(source, 0, nil)
        else {
            throw SidecarError.serverError("Failed to open image: \(path)")
        }
        return try imageData(from: cgImage)
    }

    private func loadTiff(path: String) throws -> ImageData {
        let url = URL(fileURLWithPath: path)
        guard let source = CGImageSourceCreateWithURL(url as CFURL, nil),
              let cgImage = CGImageSourceCreateImageAtIndex(source, 0, nil)
        else {
            throw SidecarError.serverError("Failed to open TIFF: \(path)")
        }
        return try imageData(from: cgImage)
    }

    private func imageData(from cgImage: CGImage) throws -> ImageData {
        let width = cgImage.width
        let height = cgImage.height
        let channels = min(cgImage.bitsPerPixel / max(cgImage.bitsPerComponent, 1), 3)
        let bytesPerRow = width * max(channels, 1)
        var pixels = [UInt8](repeating: 0, count: width * height * 3)
        let colorSpace = CGColorSpaceCreateDeviceRGB()
        let bitmapInfo = CGBitmapInfo(rawValue: CGImageAlphaInfo.noneSkipLast.rawValue)
        guard let context = CGContext(
            data: &pixels,
            width: width,
            height: height,
            bitsPerComponent: 8,
            bytesPerRow: width * 3,
            space: colorSpace,
            bitmapInfo: bitmapInfo.rawValue
        ) else {
            throw SidecarError.serverError("Failed to create image context")
        }
        context.draw(cgImage, in: CGRect(x: 0, y: 0, width: width, height: height))
        return ImageData(width: width, height: height, channels: 3, pixels: pixels)
    }
}
